File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Control/UiToolBar.ts
import { IControl } from "../UiControl";
import { AveLib } from "../../AveLib";
import { Window } from "./UiWindow";
import { CrossDirection, IconCache, DpiSize, StringKey } from "../UiCommon";
import { Vec2 } from "../../Math/Vector";

export enum ToolBarItemFlag {
    None /**/ = 0x0000,
    Id /**/ = 0x0001,
    Type /**/ = 0x0002,
    Icon /**/ = 0x0004,
    Text /**/ = 0x0010,
    Tooltip /**/ = 0x0020,
    Width /**/ = 0x0040,
}

export enum ToolBarItemType {
    Button,
    ButtonDrop,
    ButtonSplit,
    Splitter,
    Content,
    Text,
    Menu,
}

export class ToolBarItem {
    Flag: ToolBarItemFlag = ToolBarItemFlag.None;
    Id: number = 0;
    Type: ToolBarItemType = ToolBarItemType.Button;
    Icon: IconCache = 0;
    Width: DpiSize = DpiSize.FromPixel(0);
    Text: string = "";
    Tooltip: string = "";

    constructor(
        id: number = 0,
        type: ToolBarItemType = ToolBarItemType.Button,
        icon: IconCache = 0,
        width: DpiSize = DpiSize.FromPixel(0),
        text: string = "",
        tooltip: string = ""
    ) {
        this.Id = id;
        this.Type = type;
        this.Icon = icon;
        this.Width = width;
        this.Text = text;
        this.Tooltip = tooltip;
    }
}

export interface IToolBar extends IControl {
    new (window: Window, key?: StringKey): IToolBar;

    ToolInsert(item: ToolBarItem, nIndexBefore: number): boolean;
    ToolRemove(nIndex: number): boolean;
    ToolRemoveAll(): ToolBar;
    ToolGet(nIndex: number): ToolBarItem;
    ToolSet(nIndex: number, item: ToolBarItem): boolean;
    ToolGetById(nId: number): ToolBarItem;
    ToolSetById(nId: number, item: ToolBarItem): boolean;
    ToolGetCount(): number;

    ToolSetEnable(nId: number, bEnable: boolean): ToolBar;
    ToolGetEnable(nId: number): boolean;

    ToolSetVisible(nId: number, bVisible: boolean): ToolBar;
    ToolGetVisible(nId: number): boolean;

    ToolSetCheck(nId: number, bCheck: boolean): ToolBar;
    ToolGetCheck(nId: number): boolean;

    ToolHitTest(
        pt: Vec2,
        bIncludeInvisible: boolean,
        bIncludeDisabled: boolean
    ): number;

    ContentSet(nIndex: number, pControl: IControl): IControl;
    ContentGet(nIndex: number): IControl;
    ContentSetById(nId: number, pControl: IControl): IControl;
    ContentGetById(nId: number): IControl;

    DropSet(nIndex: number, pControl: IControl): IControl;
    DropGet(nIndex: number): IControl;
    DropSetById(nId: number, pControl: IControl): IControl;
    DropGetById(nId: number): IControl;

    SetBackground(b: boolean): ToolBar;
    GetBackground(): boolean;

    SetIconTextDirection(nDir: CrossDirection): ToolBar;
    GetIconTextDirection(): CrossDirection;

    OnClick(fn: (sender: ToolBar, nId: number) => void): ToolBar;
    OnDrop(fn: (sender: ToolBar, nId: number) => void): ToolBar;
}

export class ToolBar extends (AveLib.UiToolBar as IToolBar) {
    private m_Content: Set<IControl> = new Set();
    private m_Drop: Set<IControl> = new Set();

    ToolRemove(nIndex: number) {
        this.m_Content.delete(this.ContentGet(nIndex));
        this.m_Drop.delete(this.DropGet(nIndex));
        return super.ToolRemove(nIndex);
    }

    ToolRemoveAll() {
        this.m_Content.clear();
        this.m_Drop.clear();
        return super.ToolRemoveAll();
    }

    ContentSet(nIndex: number, pControl: IControl) {
        const t = super.ContentSet(nIndex, pControl);
        if (t) this.m_Content.delete(t);
        if (pControl) this.m_Content.add(pControl);
        return t;
    }

    ContentSetById(nId: number, pControl: IControl) {
        const t = super.ContentSetById(nId, pControl);
        if (t) this.m_Content.delete(t);
        if (pControl) this.m_Content.add(pControl);
        return t;
    }

    DropSet(nIndex: number, pControl: IControl) {
        const t = super.DropSet(nIndex, pControl);
        if (t) this.m_Drop.delete(t);
        if (pControl) this.m_Drop.add(pControl);
        return t;
    }

    DropSetById(nId: number, pControl: IControl) {
        const t = super.DropSetById(nId, pControl);
        if (t) this.m_Drop.delete(t);
        if (pControl) this.m_Drop.add(pControl);
        return t;
    }
}

File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiToolBar.cpp
#include "StdAfx.h"
#include "UiToolBar.h"

#define ThisMethod($x) &UiToolBar::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiToolBar::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiToolBar> c_obj;
	}

	void WrapDataTransit<Ui::ToolBarItem>::Transit( Ui::ToolBarItem& d, const UiToolBarItem& s )
	{
		d.m_Flag    /**/ = s.m_Flag    /**/;
		d.m_Id      /**/ = s.m_Id      /**/;
		d.m_Type    /**/ = s.m_Type    /**/;
		d.m_Icon    /**/ = s.m_Icon    /**/;
		d.m_Width   /**/ = s.m_Width   /**/;
		d.m_Text    /**/ = (PWChar) s.m_Text    /**/.c_str();
		d.m_Tooltip /**/ = (PWChar) s.m_Tooltip /**/.c_str();
	}

	void UiToolBar::DefineControl()
	{
		AutoAddMethod( ToolInsert );
		AutoAddMethod( ToolRemove );
		AutoAddMethod( ToolRemoveAll );
		AutoAddMethod( ToolGet, WrapObjectGeneric );
		AutoAddMethod( ToolSet );
		AutoAddMethod( ToolGetById, WrapObjectGeneric );
		AutoAddMethod( ToolSetById );
		AutoAddMethod( ToolGetCount, WrapObjectGeneric );
		AutoAddMethod( ToolSetEnable );
		AutoAddMethod( ToolGetEnable, WrapObjectGeneric );
		AutoAddMethod( ToolSetVisible );
		AutoAddMethod( ToolGetVisible, WrapObjectGeneric );
		AutoAddMethod( ToolSetCheck );
		AutoAddMethod( ToolGetCheck, WrapObjectGeneric );
		AutoAddMethod( ToolHitTest, WrapObjectGeneric );
		AutoAddMethod( ContentSet );
		AutoAddMethod( ContentGet, WrapObjectGeneric );
		AutoAddMethod( ContentSetById );
		AutoAddMethod( ContentGetById, WrapObjectGeneric );
		AutoAddMethod( DropSet );
		AutoAddMethod( DropGet, WrapObjectGeneric );
		AutoAddMethod( DropSetById );
		AutoAddMethod( DropGetById, WrapObjectGeneric );
		AutoAddMethod( SetBackground );
		AutoAddMethod( GetBackground, WrapObjectGeneric );
		AutoAddMethod( SetIconTextDirection );
		AutoAddMethod( GetIconTextDirection, WrapObjectGeneric );
		AutoAddMethod( OnClick, WrapObjectGeneric );
		AutoAddMethod( OnDrop, WrapObjectGeneric );
	}

	U1 UiToolBar::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		GetControlTyped().SetIconManager( p->PublicCloneIconManager() );

		return true;
	}

	void UiToolBar::__OnClick( Ui::IToolBar & sender, const Ui::ToolBarItem & item )
	{
		m_OnClick( this, item.m_Id );
	}

	void UiToolBar::__OnDrop( Ui::IToolBar & sender, const Ui::ToolBarItem & item )
	{
		m_OnDrop( this, item.m_Id );
	}

	U1 UiToolBar::ToolInsert( const WrapData<UiToolBarItem>& item, U32 nIndexBefore )
	{
		Ui::ToolBarItem ti{};
		WrapDataTransit<Ui::ToolBarItem>::Transit( ti, item );
		return GetControlTyped().ToolInsert( ti, nIndexBefore );
	}

	WrapData<UiToolBarItem> UiToolBar::ToolGet( U32 nIndex ) const
	{
		Ui::ToolBarItem ti{ Ui::ToolBarItem::Flag_Id | Ui::ToolBarItem::Flag_Type | Ui::ToolBarItem::Flag_Icon | Ui::ToolBarItem::Flag_Text | Ui::ToolBarItem::Flag_Tooltip | Ui::ToolBarItem::Flag_Width };
		GetControlTyped().ToolGet( nIndex, ti );
		WrapData<UiToolBarItem> r{};
		WrapDataTransit<Ui::ToolBarItem>::Transit( r, ti );
		return r;
	}

	U1 UiToolBar::ToolSet( U32 nIndex, const WrapData<UiToolBarItem>& item )
	{
		Ui::ToolBarItem ti{};
		WrapDataTransit<Ui::ToolBarItem>::Transit( ti, item );
		return GetControlTyped().ToolSet( nIndex, ti );
	}

	WrapData<UiToolBarItem> UiToolBar::ToolGetById( U32 nId ) const
	{
		Ui::ToolBarItem ti{ Ui::ToolBarItem::Flag_Id | Ui::ToolBarItem::Flag_Type | Ui::ToolBarItem::Flag_Icon | Ui::ToolBarItem::Flag_Text | Ui::ToolBarItem::Flag_Tooltip | Ui::ToolBarItem::Flag_Width };
		GetControlTyped().ToolGetById( nId, ti );
		WrapData<UiToolBarItem> r{};
		WrapDataTransit<Ui::ToolBarItem>::Transit( r, ti );
		return r;
	}

	U1 UiToolBar::ToolSetById( U32 nId, const WrapData<UiToolBarItem>& item )
	{
		Ui::ToolBarItem ti{};
		WrapDataTransit<Ui::ToolBarItem>::Transit( ti, item );
		return GetControlTyped().ToolSetById( nId, ti );
	}

	UiToolBar* UiToolBar::OnClick /**/( Callback_t&& fn ) { SetEventCallback<Ui::IToolBar::OnClick /**/>( m_OnClick /**/, std::move( fn ), MakeThisFunc( __OnClick /**/ ) ); return this; }
	UiToolBar* UiToolBar::OnDrop  /**/( Callback_t&& fn ) { SetEventCallback<Ui::IToolBar::OnDrop  /**/>( m_OnDrop  /**/, std::move( fn ), MakeThisFunc( __OnDrop  /**/ ) ); return this; }
}
